src/Middleware/Localization/src/RequestLocalizationMiddleware.cs

```diff
@@ -163,67 +162,73 @@ private static void SetCurrentThreadCulture(RequestCulture requestCulture)
+
+                if (cultureInfo == null)
                 {
-                    var cultureInfo = GetCultureInfo(cultureName, supportedCultures, fallbackToParentCultures, currentDepth: 0);
-                    if (cultureInfo != null)
+                    if (fallbackToParentCultures)
                     {
-                        return cultureInfo;
+                        var fallbackCulture = CultureInfo.GetCultureInfo(cultureName.Value);
+
+                        while (fallbackCulture != fallbackCulture.Parent)
+                        {
+                            fallbackCulture = fallbackCulture.Parent;
+
+                            if (supportedCultures.Contains(fallbackCulture))
+                            {
+                                return fallbackCulture;
+                            }
+                        }
+
+                        if (supportedCultures.Contains(fallbackCulture))
+                        {
+                            return fallbackCulture;
+                        }
                     }
                 }
+                else
+                {
+                    return cultureInfo;
+                }
             }
 
             return null;
-        }
 
-        private static CultureInfo GetCultureInfo(StringSegment name, IList<CultureInfo> supportedCultures)
-        {
-            // Allow only known culture names as this API is called with input from users (HTTP requests) and
-            // creating CultureInfo objects is expensive and we don't want it to throw either.
-            if (name == null || supportedCultures == null)
+            IEnumerable<StringSegment> GetValidNonNullableCultures()
             {
-                return null;
-            }
-            var culture = supportedCultures.FirstOrDefault(
-                supportedCulture => StringSegment.Equals(supportedCulture.Name, name, StringComparison.OrdinalIgnoreCase));
+                var validCultures = new List<StringSegment>();
 
-            if (culture == null)
-            {
-                return null;
-            }
-
-            return CultureInfo.ReadOnly(culture);
-        }
-
-        private static CultureInfo GetCultureInfo(
-            StringSegment cultureName,
-            IList<CultureInfo> supportedCultures,
-            bool fallbackToParentCultures,
-            int currentDepth)
-        {
-            var culture = GetCultureInfo(cultureName, supportedCultures);
+                foreach (var name in cultureNames)
+                {
+                    try
+                    {
+                        if (name == null)
+                        {
+                            continue;
+                        }
 
-            if (culture == null && fallbackToParentCultures && currentDepth < MaxCultureFallbackDepth)
-            {
-                var lastIndexOfHyphen = cultureName.LastIndexOf('-');
+                        var culture = CultureInfo.GetCultureInfo(name.Value);
```

Comment:
This is still creating a CultureInfo on every request. Check out [this PR](https://github.com/aspnet/AspNetCore/pull/14550). If there's an acceptable solution to this problem I think it probably looks something like that, because there's a finite one-time cost. That said, there's still significant risk that it's too non-performant. I'll re-open this PR but please keep the CultureInfo.Parent issue separated from the `ResourceManager.GetString()` thing so that everything is easy to follow.

Reply:
> This is still creating a CultureInfo on every request. 

Let me check it again

Reply:
> This is still creating a CultureInfo on every request. 

`CultureInfo.GetCultureInfo()` retrieves a cached, read-only instance of a culture.

Reply:
(I'll assume here that you're referring to https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Globalization/CultureData.cs#L536 when you talk about the caching). Caching user data doesn't really do anything for us if we're concerned about malicious requests (this is user data, so we should be). They can easily do Culture=aa-AA UICulture=aa-AB, Culture=aaa-AAA UICulture=aaa-AAB, so on and so forth. The difference with the other PR is that they only create instances from the systems (finite) list of "known good" cultures, so they're not doing the expensive operation on (effectively infinite) user data.

Reply:
Caching all known cultures are good and bad, good because it will do it once and cache forever, and bad because it may affect the performance on the first request, I come up with another option is to define something like `MaxCulturesLimit` with the default value `5`, so whatever the browser it sends it will only take the first `5` culture which is fine instead of all the known cultures

If you remember will already had a list of the system cultures from DNX days, after that we decided to substituted with the supported cultures

If sound is good I can modify the PR and add a unit tests for it

Reply:
/cc @ryanbrandenburg @pranavkm